### DISCUSSION TITLE
Circuit Enclosure (Circuit Boards / Chips)

### DISCUSSION THREAD

**Discussion by Saevon** [FakeFishGames/Barotrauma#4721] [Ideas and suggestions]
# Problem

Currently any sort of logic is impossible to interfere with for crew in-game (Anything over ~2 components). Most sub-creators simply overlap every single component into the walls once they're done wiring. This means sub-editing is a pain afterwards, and its impossible to edit in-game.

This also means sub-creators spend hours just getting a system from a workshop working in their ship.

# Solution

There's no real reason the circuits NEED to be physically on the sub. If instead there could be "Circuit Enclosures". These are 1x1 tile items, which have 2 wiring modes. `[E] Rewire (Screwdriver)` would show the wires view, `[Click] Edit Circuit (Screwdriver)` would show the circuit view

## Example Circuit

This is a circuit that uses hardcoded "hull-size" values (in memory) as its input, Inputting the sum of `flood % * Hull-size` for each room.

Adjusts speed to account for how flooded the ship is. Has a stability toggle that prevents motion over a certain speed.

### Wires View

<img width="660" alt="Circuit-1-Wire" src="https://user-images.githubusercontent.com/570036/104261005-72c03400-5439-11eb-923e-92095cb2466a.png">

### Circuit View

![Circuit-1-Circuit](https://user-images.githubusercontent.com/570036/104261200-de0a0600-5439-11eb-969e-50d7c8005c09.png)

## Example 2

This shows how some non-components would be useful (a switch) allowing a circuit to be slightly customizable for re-usage (or change-able in game, with circumstances)

This is the input part of the flood circuit, requiring you to type in the hull-size (mem) of the current room. Then outputs the square meter flooded. Can accept a "is this ballast" control (in which case ignore the "flood") which can be controlled through the switch (so a sub-editor can enable/disable it permanently) or an input, so people can automate backup-ballast.

<img width="1197" alt="Circuit-3" src="https://user-images.githubusercontent.com/570036/104261754-18c06e00-543b-11eb-9bd8-bdf70760ce07.png">

Other useful items:

 * Buttons / Switches / Levers
 * Text / Lights, allowing engineers to open a circuit enclosure to see immediate debug info
 * Labels

# Additional Info

Ideally you could select the Circuit Enclosure and "Save as Item Assembly" then these could be used to create circuit mods. 

 * Want a smart airlock?   Here's a chip you can download that you just need to wire out the buttons / pump / doors to.
 * Want a smart OS?   Here's the features it supports, and the connections you need to make.

For simpler circuits, this item could allow rewiring mid-game, as now nothing is hidden behind walls. Or at least unintelligent things could be rewired on the spot.

Some of these chips could also be dropped as an item. Getting some a "flood seal" chips could be a reward for some submarines (locks ducts when section is over 20% water, or lower section is flooded)

<details>
  <summary><strong>Community Examples</strong></summary>

Complicated label usage for a reactor circuit (whose info will be lost)
![unknown-12](https://user-images.githubusercontent.com/570036/104836631-1196c700-5864-11eb-9038-32c16ea1d0d0.png)

Example output + labels used for explanation (also reactor circuit)
![images-ext-2 discordapp](https://user-images.githubusercontent.com/570036/104836633-1491b780-5864-11eb-9831-7fa505c7caf1.jpg)

</details>

**Comment by Saevon** [comment]
P.S. If these could also be nested… that would be super awesome!

**Reply by JamesTCrenshaw** [reply]
I like it. I feel like this could be a good solution for things like doors and such while still allowing players to edit them without making a visible and ugly birdnest of wires. I'm working on a couple subs and I don't like the idea of having to hide all the wiring because I'd rather players be able to edit it should they want to - but at the cost of it being fairly ugly or taking up a rather large amount of space so that people can understand what each component is doing.

**Comment by rav2n** [comment]
Personally, I'm against it.

The first reason is ruiners and inexperienced players. There are a number of submarines in the logical circuits of which it is better not to climb during the game. You can ban a player of course, but what to do with damaged electronics? Electronics are hidden within the walls not from a good life, but so that they are not accidentally or intentionally broken. If you give access to logic circuits during the game, then only at outposts, with a preliminary opportunity to save before that.

The second reason, I think this is not fair in relation to those who invented these electronic circuits. The approach you have proposed will in part devalue their work. I'm not against someone copying interesting solutions. I am against thoughtless copying. Your proposed solution makes subbuilders less motivated to learn about electronics within the game. Consequently, you are depriving players of a huge layer of game content. This approach will only widen the gap between the knowers and those who don't know. For all other cases, it is possible to upload item assemblies to the workshop and communicate between players.

**Reply by Saevon** [reply]
> Electronics are hidden within the walls not from a good life, but so that they are not accidentally or intentionally broken.

There is no reason it cannot be both, interactable is still a flag, which could still be applied to both the inside components and to the circuit itself if needed. Currently there are many easily breakable subs, you can grab wires and break things as is: we have to mark specific doors, circuits, etc as untouchable for this very reason. (You can still hide enclosures in walls as well, nothing stops you there)

<hr>

> If you give access to logic circuits during the game, then only at outposts, with a preliminary opportunity to save before that.

Great idea!  I've actually wanted to be able to "buy parts" between missions, effectively re-building parts of the sub between missions. Allowing players to slowly customize a vanilla sub during campaign could be a way to get more crew-players into sub-builders.

<hr>

>  I think this is not fair in relation to those who invented these electronic circuits. The approach you have proposed will in part devalue their work.

Not sure I agree?  I make tons of circuits in various games, I would feel quite happy to have my work used and modified, helping others in their own circuit quests when I can. Others of course would not agree, wanting full recognition, and to prevent "slight changes" from being a new work.

<details>
  <summary>This is simply the free-source vs open-source debate</summary>

**_When we call software “free,” we mean that it respects the users' essential freedoms: the freedom to run it, to study and change it, and to redistribute copies with or without changes. This is a matter of freedom, not price, so think of “free speech,” not “free beer.”"_**
</details>

I lean one way, you lean the other.

To answer:  I've seen three types of circuits

 * Purposefully spread out and easy to understand, sometimes with helpers (but with a second compressed item-assembly)
 * Compressed 1x1 item-assembly that has hidden wires, super miniaturized components, and other compression-obfuscation attempts.
 * One's in a sub which don't expect an individual circuit to be pulled out.

The first two approaches are still 100% supported.  The first no longer needs to have a sample with decompressed AND compressed. The second simply compresses INSIDE the Circuit enclosure.

Instead both gain the input output labelling that is already being used by both. (Text labels and arrows in the item-assembly used to indicate hook up points).

The third is more complicated, I do not know it they want their sub to be proprietary, (considering you can already download and change it easily). Or if their circuits are just messy for lack of an alternative.

<hr>

> Your proposed solution makes subbuilders less motivated to learn about electronics within the game. Consequently, you are depriving players of a huge layer of game content.

This is a problem, thank you for pointing it out!

One of my aims is to help newer players get into circuits!   There is great power to creating an entire circuit from scratch, but there is also power from grabbing a circuit you like, then modifying it to also detect fires, or send funny inside jokes, or what have you.

The main problem here is people releasing "optimal" chips, and yeah I don't really want this either…  not really a solution here right now. Perhaps the community has more ideas, or the devs? 

<hr>

The other advantage here is mid-ship wiring, I know coop crews that rewire their batteries as they run the humpback for their own purposes, who would love to be able to add their own logic without worrying about space on the ship (because once the ship is built you're NEVER going to have room for circuits and wires anymore).

 * As a new player if there was a "wiring kit" (any 10 components/wires: so you don't have to worry about the differences), and an enclosure (No need to worry about space, or polluting the ship with experiments)… adding a quick "warn me when the reactor is really hot" or "make noise if this forgotten corner is flooding" might now be doable… if you can now get into the circuits without being a "sub builder", this can encourage you to actually start modifying the ship!   Crew-players might now become sub-builders.

Instead as a professional programmer, I almost quit sub-building knowing that anything more then 3 components was going to make moving stuff around impossible… I would need to build a perfect sub the _very first_ time.  Even now I'm hesitant to mess around with any circuits unless I've perfectly tested and build it elsewhere (knowing that It'll be impossible to debug/fix/improve properly if I compress it)



TL;DR
 * Yes, Optimal chips are a problem for creativity
 * We disagree on open/free source principles
 * Hopefully this _can_ help new players!

**Reply by rav2n** [reply]
> Not sure I agree? I make tons of circuits in various games, I would feel quite happy to have my work used and modified, helping others in their own circuit quests when I can. Others of course would not agree, wanting full recognition, and to prevent "slight changes" from being a new work.

As I already said. I am calm about the fact that someone is copying and modifying my works. But there are players who are not satisfied with this approach, I have witnessed such situations with submarines more than once. Sometimes it turned into unpleasant skirmishes.

**Reply by Saevon** [reply]
An addition to the anti-griefing:    Currently you can only "lock" a component or wires.  With this we'd want to be able to lock the wiring-panel or the Circuit-panel separately from the item…

 * non-Interactable would be for things we cannot affect at all.
* Repairable would be for things that we can interact to repair
* Wireable would be for things we can edit the wiring-panel for
* Circuitable would be for things we can edit the circuit-panel for
* Circuit-addable would be if we can add things inside the panel (Useful if you have switches / lights inside the circuits that you want people to be able to view...  (Alternatively allow buttons/switches/lights to appear on the wire-panel similar to how the terminal copper-circles can be added)

**Reply by Saevon** [reply]
Relevant Link:  [Devs: "Plagiarism is frowned upon"](https://steamcommunity.com/workshop/discussions/18446744073709551615/1751276319483713708/?appid=602960)

**Reply by FalcoGer** [reply]
> Personally, I'm against it.
> The first reason is ruiners and inexperienced players.

If anything it makes griefing harder, because now they have to mess with macro circuits that are much more easy to see what does what and what is supposed to go where. Assholes will be assholes, if you take away their screwdriver they will just throw your supplies overboard instead. Just ban them and be done with it. Additionally, one may be able to lock the components and wires inside such black boxes entirely, allowing circuits to be hooked up outside of them during the round, but not mess with the prebuilt stuff, besides unhooking the external connections, which again, are much easier to fix than someone undoing a tangled mess of components that was impossible to understand even while it had been wired up properly.

The same goes for inexperienced players. They see a tangle of wires and 30 components stuffed into a tangled mess and they think: "Let's not mess with that". If they do mess with it, they just see a forest of components connected, seemingly at random. When I design a new circuit, I lay it out spaciously and logically, put labels everywhere, write down formulas and try to make it as easy for me to remember what I was thinking when I first made the thing. Because once you put the whole thing into a wall inside your submarine, the whole neatness and logic just disappears and now you just don't know what's doing what anymore, unless it's something simple. Compare that with nice, big black boxes, that when opened give you the same spacious, commented layout as was in the circuit designer's head when he made the thing. Suddenly things will be a lot more clear to everyone involved. And from the outside you get high level components connected together, again making things easier to both understand and to fix if some idiot were to unwire it all.

> The second reason, I think this is not fair in relation to those who invented these electronic circuits.

I don't get it. You can download assemblies off the workshop and copy paste all you want already. Or you can develop your own stuff just the same. If people put stuff for other people to use, then they are obviously fine with it.

It literally doesn't change a thing, except it makes everything more tidy.

**Reply by FalcoGer** [reply]
> Yes, Optimal chips are a problem for creativity

Nonsense. Designing optimal chips is great fun. I had a whole assembly of reactors with different outputs and batteries with different loads and I took a whole day to learn the formulas for

- Setting a reactor to whatever temperature I want
- Getting the minimum temperature for a given output (and combine with above)
- Getting the turbine setting for a given load

along with plenty of other interesting things about reactors that wasn't described on the wiki at all.

The math is straight forward enough after you figure it out and the resulting wire job is perfect. If you do a bit of math you can do a 5 component job to keep the minimum temperature.

So designing the thing was good fun. You claim however that this "perfect circuit" would be a bad thing. Why is that? Not using it means that you use more fuel, and perhaps not get a great turbine response to load changes. That's literally it. What extra fun is gained by having to buy or fabricate another fuel rod every 3 rounds? Compare that with the 2 component bang bang jobbie that just feels like cheating and it isn't optimal either since it just keeps the temperature at some value. Or perhaps you have more fun with no reactor wires at all, relying on the automation that blows all your junction boxes up as soon as you stop the engine?

Maybe having perfect chips on the sub makes wireing on the go obsolete, but that wasn't the argument. If such chips aren't on the sub yet, you can just make them on the go and hook them in, improving your sub and thus "being fun".

**Reply by Saevon** [reply]
Wanted to add btw, this item can ALSO be placed behind the wall and made non-interactable SAME as existing logic!   So it offers no downsides but cleanliness.

So in that case a sub editor gets cleaner, debugable, reusable circuits,,, even while hiding it if they want

**Comment by Working-Joe** [comment]
I am much in favour of this. The current wiring system can get very cluttered, being able to place circuit boxes would improve the looks of subs. 9 out of 10 times they're hidden in the walls now anyways. 

Griefing doesn't have to be a problem, since it's already possible to disable wiring, although I would advise against it. Being able to lock boxes to just engineers would be an interesting alternative. 

Of course balancing would need to be taken into acount in the campaign to prevent people from greatly improving their subs with a single item. Just a few things that come to mind that might work:
-A power requirement depending on the number of contained components. 
-different sizes of circuit boxes with limits on the number of inputs/outputs and internal components. 

Although I share the sentiment that learning electronics is a fun part of the sub building experience, it's already possible to share circuits on the workshop. But people often end up wanting stuff that's a bit more customised to their subs. On the flipside, wiring is a huge hurdle towards sub builders as it's really hard to get into. Being able to get premade circuit boxes would greatly help newer players, especially if a few premade ones are added. 

If you still want the average sub-enjoyer to learn wiring, it might be possible to add more complicated wiring puzzles to beacons at some point :)

**Reply by Saevon** [reply]
I love the idea of wiring puzzles!

**Comment by mLuby** [comment]
For anti-griefing, could add an in-game "padlock" item to the circuit enclosure that takes a long time to pick. Even a more advanced "alarmed padlock" could report any tampering to the chat.

Even a circuit room with a motion sensor could do this (less elegantly).

**Reply by Saevon** [reply]
Padlocks are an awesome idea!!   even on normal items this could be so cool.   You can padlock the security door, and thats how the "id_security" would work… but now you can pick it with alarms blaring

In general there could be important junctions that are now "Traitor protected" (but not impossible)

**Comment by Ek-Videogames** [comment]
Personally not a fan of the alternative we have right now which is people often not even trying to organize their wiring and instead just setting everything to non-interactable and hidden-ingame. Thus resulting in the same problem of not being able to access wiring.

**Reply by ubertpendragon** [reply]
Not being able to access wiring in subs makes sense because there are only two reasons to do "live" wiring right now:
-Griefing
-Building auto-reactor

I disabled wiring on my sub because there's no room for improvement, whereas there's always the opportunity to grief
 Beacons and wrecks are where electrical engineers should flex, so I'd encourage more complex beacon generation instead

**Reply by mLuby** [reply]
> Not being able to access wiring in subs makes sense because there are only two reasons to do "live" wiring right now:
> -Griefing
> -Building auto-reactor

Respectfully, that's a somewhat limited view. Wiring projects I've enjoyed in-game:

- [x] Turn a docking-only hatch into an improvised second airlock.
- [x] Wiring doors to seal when flooded, open when dry, and respect manual interactions.
- [x] Improve airlock wiring to prevent accidentally flooding the rest of the sub.
- [x] Add lighting to cheer up gloomy compartments.
- [x] Prepare ballast pumps for anti-flora action.
- [x] Add enemy/O2/water detection and indicators/alarms.

(Edits due to weird formatting.)

**Comment by NilanthAnimosus** [comment]
If this system could be done in game visually without restriction (IE. attach a circuit enclosure to a wall, and you can add/remove and rewire components inside the enclosure that other players can see you modifying in-game) I am all for this, As it is very commonplace for key logical wiring to be simply disabled from rewiring or modifying and offloaded into the walls. It also takes a lot of space, and many unnecessarily drawn wires and components you cant see.

This does add a choice of taking time to edit the circuits (Like ID-access and crowbaring to modify even the external wiring of it) - Visibility it has been secured or unsecured (IE. a hanging off panel sprite), and removes the 2-5 second cable tear-out or rearrange issue people will typically disable rewiring for. I could never picture it being reasonable to individually-secure singular logic components without ensuring insanity from a player.

To boot if I see an airlock in game with 20 wires going in 20 different directions all crossed over between themselves amongst 5-9 components and someone removes a cable? I am just connecting those buttons and ripping the logic out. A neatly done enclosure though just wired to the doors with a less compacted easier to work with view? sounds too good to be true.

Perhaps it could be detailed in-game what is meant to go where - such as a sub designers snapshot of the original components (Or generated on round load) and layout as a background or other such, an in game generated diagram conceptually using a simple texture with half/quarter alpha (This could even be done at round load after save load but before sync up). Whatever is done with this feature if its ever a thing I think it would be cool to have.

Personally I would use a set number of input/output pins, have them labelable, then everything else be free real estate. with the interface being somewhat larger than the standard rewiring panel and centered, perhaps fading if looking to the sides of the screen.

Between new concepts for the security/griefing side to allow for a fair time restriction to rewire, more space to work with in game in component layout, and less directly-in-world clutter. This seems nice overall as a feature sub designers could optionally use.

**Comment by Kulik242** [comment]
Engineer (or person with high enough electrical skill), should be able to see the original setup of the circuit board as a ghost image - seeing what changes were made to the original design, they should also be able to reset to the original values inside components (output, false output, signals, etc.).

This would make even unskilled player be able to easily fix changes made by trolls, griefers and traitors.

Outpost could provide option to repair all circuits to their original state.

**Reply by mLuby** [reply]
Good ideas; you may want to add them to the [Wiring repair at Station](https://github.com/Regalis11/Barotrauma/discussions/4743) discussion on here.

**Comment by Hinny** [comment]
The more I think about it the better this suggestion is!

This would be such a great addition to the engineering gamplay and a huge quality of life improvement for both sub builders and in-game tinkerers. It would also make it easier to grasp for beginners (better overview, more managble "bite size" and in-game help if everything is labeled correctly etc). Not to speak about the visual improvement this would bring to the subs (naked components are ugly, that is one reason why almost everyone hides them). I would argue this would be the best improvement you can do for the game right now. 

You can even make the circut boards deteriorate similar to a junction box. Also make them susceptible to water, emp and fire as they should be. If it is damaged the output could be unreliable or simply just no output at all. This would also make a high-tech sub less robust and more high maintenance as it should be in my opinion. 

I like the suggestion with editable input / output labels (maybe 4 of each) and a fixed grid for components (maybe 16 x 16) and maybe an editable title label as well. If you want more complicated functionality you could have several set up in series so that would not be a problem.

The question is if these would replace the old system of placing naked components on the walls? I lean towards replacing completely. If you just want have a couple of wifi components or similar simple set up it is still doable --- you could even combine up to four wifi components in each circuit board (each connected to a single output; inputs not connected)

Please devs, make circuit boards happen! 🙂

**Comment by Hinny** [comment]
I like the idea of padlocks as well, but that would be a separate feature in my opinion (you could maybe even use this to lock junction boxes and cabinets as well). Combination code padlocks would probably be the easiest to implement. A crowbar should probably be able to destroy them given time.

**Comment by Leonidas-ai** [comment]
This sounds l like an idea that should have always been included in the game. Some kind of framework where you can set inputs and outputs and then do crazy stuff with components inbetween and then embed it into your sub. I guess it probably should have been a high level design decision long before this point so it may not be feasible anymore.

**Comment by HeatedWafflez** [comment]
I really like this idea since it allows players to customize submarines in game to their liking. If griefing is a problem, then people should still be able to place logic components outside of circuit boards whenever they like in case they intend to hide them behind walls. This is so that way sub builders can designate which logic components are to be changed and which aren't.

**Comment by MaaiG** [comment]
The more components I have in my circuits, the more I want to be able to assemble them into one integrated circuit.

This will allow:
- reduce the space that the circuits occupy
- simplify work with them when they are placed
- will make the work with components contained in them more compact and convenient.
![Безымянный-1](https://user-images.githubusercontent.com/93348499/173801183-e53aa3e2-b3e2-4122-8542-24f97f4f31b6.png)

P.S. I like that too https://github.com/Regalis11/Barotrauma/discussions/5001

**Comment by FalcoGer** [comment]
I like the idea. I find myself copying circuits of 4-16 components all over my ship over and over again to add such basic functionality such as automated doors, hatches (which use particularly many components since they have to be automatic for both people and water and i added some convenient features as well), or airlocks or a central alarm system or a minimum component that spits out the lower value of two inputs. It would be really great if you could all just group such a thing into one component, save the space, place them where they make sense and most of all save myself the work of copy&pasting large chunks and then having to remember what's what in which circuit.

What's more, this could allow to update all the components at once, at least in the submarine builder. I want to add a feature or fix a bug, i now have to go back and either manually rewire every individual circuit, or delete them, fix up the floating wires that are left behind to the connections all over the ship, remember where those were originally connected to, paste in the new, fixed version and then rewire it to the rest of the ship without missing anything. It's just a big pain to deal with and I sunk plenty of hours into such nonsense. If I could just have all the components be in a single box that could be updated for everywhere it occurs, given that the connections to the outside don't change, then sub making would be a lot faster.

**Reply by NaturalBornCamper** [reply]
It would be a big advantage in the base have as well, without using custom subs.

It would be nice to make a circuit for reactor control, save it as a "chip" with our inputs and outputs, then when you transfer to a new sub you don't have to re-create everything and make mistakes along the way, loading precious game time. You would just detach your chip and move it to the new sub.

The game could even count the resources in your chip then when going to the fabricator, you can make a chip there, for example a motion-sensor chip for doors. Too bad the fabricator only has room for 5 ingredients though, this would cause problems if a chip has more than 5 components.

**Comment by egpimp** [comment]
Having to hide my entire circuits just for the sake of making the sub look clean is not ideal, fully agree with this

**Comment by Regalis11** [comment]
Implemented
